### PR TITLE
weight map mask for boundaries

### PR DESCRIPTION
**Pull request by pranathivemuri**
image

calculating distance map according to this equation in unet paper

maybe needs more docs, but the mask part is done and its working

here is how it looks like when 2 cells are close then the distance map's weights higher at those points
image

### CODE REVIEW

**Comment by pranathivemuri** [comment]
@anithapk please review when you are free!

**Comment by pranathivemuri** [comment]
@anithapk please review when you are free!

File: micro_dl/utils/masks.py [hunk 1]
```diff
@@ -94,3 +95,65 @@ def create_unimodal_mask(input_image, str_elem_size=3):
     thr_image = binary_opening(input_image > thr, str_elem)
     mask = binary_erosion(thr_image, str_elem)
     return mask
+
+
+def get_unet_border_weight_map(annotation, w0=10, sigma=5):
+    """
+    Generate the weight map for borders as specified in the UNet paper for an annotation.
+    Parameters
```
Comment (anithapk) [inline micro_dl/utils/masks.py]: Could you please make the docstrings similar to the ones we have in the repo? On hindsight numpy style would've been better...

File: micro_dl/utils/masks.py [hunk 2]
```diff
@@ -94,3 +95,65 @@ def create_unimodal_mask(input_image, str_elem_size=3):
...
+    Returns
+    -------
+    array-like
+        A 2D array of shape (image_height, image_width)
+
+    TODO: Calculate boundaries directly and calculate distance from boundary of cells to another
+    TODO: The below method only works for UNet Segmentation only
+    """
+
+    assert annotation.dtype == np.uint8, "datatype expected uint8, it is {}".format(annotation.dtype)
```
Comment (anithapk) [inline micro_dl/utils/masks.py]: Please break the line at 80 characters, in this case the error message can go on a separate line

File: micro_dl/utils/masks.py [hunk 3]
```diff
@@ -94,3 +95,65 @@ def create_unimodal_mask(input_image, str_elem_size=3):
...
+    # class balance weights w_c(x)
+    unique_values = np.unique(annotation).tolist()
+    weight_map = [0] * len(unique_values)
+    for index, unique_value in enumerate(unique_values):
+        mask = np.zeros((annotation.shape[0], annotation.shape[1]), dtype=np.float64)
+        mask[mask == unique_value] = 1
+        if mask.sum() != 0:
```
Comment (anithapk) [inline micro_dl/utils/masks.py]: Not sure if u need an if-else here as unique_values are the unique values in the annotation and the count will never be zero..

File: micro_dl/utils/masks.py [hunk 4]
```diff
@@ -94,3 +95,65 @@ def create_unimodal_mask(input_image, str_elem_size=3):
...
+
+    wc = np.zeros((mask.shape[0], mask.shape[1]), dtype=np.float64)
+    for index, unique_value in enumerate(unique_values):
+        wc[mask == unique_value] = weight_map[index]
+
+    # cells instances for distance computation
+    labeled_array, _ = scipy.ndimage.measurements.label(annotation)
+    # cells distance map
+    border_loss_map = np.zeros((annotation.shape[0], annotation.shape[1]), dtype=np.float64)
+    distance_maps = np.zeros((annotation.shape[0], annotation.shape[1], np.max(labeled_array)), dtype=np.float64)
```
Comment (anithapk) [inline micro_dl/utils/masks.py]: Please break lines that have > 80 characters

File: micro_dl/utils/masks.py [hunk 5]
```diff
@@ -94,3 +95,65 @@ def create_unimodal_mask(input_image, str_elem_size=3):
...
+    if np.max(labeled_array) >= 2:
+        for index, label in enumerate(range(1, np.max(labeled_array) + 1)):
```
Comment (anithapk) [inline micro_dl/utils/masks.py]: just a thought, simplify may be, not combine enumerate and range, for index in range(np.max(labeled_array): label = index + 1

File: micro_dl/utils/masks.py [hunk 6]
```diff
@@ -94,3 +95,65 @@ def create_unimodal_mask(input_image, str_elem_size=3):
...
+    border_loss_map = w0 * np.exp((-1 * (d1 + d2) ** 2) / (2 * (sigma ** 2)))
+
+    zero_label = np.zeros((annotation.shape[0], annotation.shape[1]), dtype=np.float64)
+    zero_label[labeled_array == 0] = 1
+    border_loss_map = np.multiply(border_loss_map, zero_label)
+
+    # unet weight map mask
+    weight_map_mask = wc + border_loss_map
+
+    return weight_map_mask
```
Comment (anithapk) [inline micro_dl/utils/masks.py]: 💯

**Comment by jennyfolkesson** [comment]
@pranathivemuri nice work!

**Pull request merged**